TransactionProcessor.BusinessLogic/EventHandling/TransactionDomainEventHandler.cs

```diff
@@ -0,0 +1,184 @@
+                                 CancellationToken cancellationToken)
+        {
+            await this.HandleSpecificDomainEvent((dynamic)domainEvent, cancellationToken);
+        }
+
+        /// <summary>
+        /// Gets the token.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        /// <returns></returns>
+        [ExcludeFromCodeCoverage]
+        private async Task<TokenResponse> GetToken(CancellationToken cancellationToken)
+        {
+            // Get a token to talk to the estate service
+            String clientId = ConfigurationReader.GetValue("AppSettings", "ClientId");
+            String clientSecret = ConfigurationReader.GetValue("AppSettings", "ClientSecret");
+            Logger.LogInformation($"Client Id is {clientId}");
+            Logger.LogInformation($"Client Secret is {clientSecret}");
+
+            if (this.TokenResponse == null)
+            {
+                TokenResponse token = await this.SecurityServiceClient.GetToken(clientId, clientSecret, cancellationToken);
+                Logger.LogInformation($"Token is {token.AccessToken}");
+                return token;
+            }
+
+            if (this.TokenResponse.Expires.UtcDateTime.Subtract(DateTime.UtcNow) < TimeSpan.FromMinutes(2))
+            {
+                Logger.LogInformation($"Token is about to expire at {this.TokenResponse.Expires.DateTime:O}");
+                TokenResponse token = await this.SecurityServiceClient.GetToken(clientId, clientSecret, cancellationToken);
+                Logger.LogInformation($"Token is {token.AccessToken}");
+                return token;
+            }
+
+            return this.TokenResponse;
+        }
+
+        /// <summary>
+        /// Handles the specific domain event.
+        /// </summary>
+        /// <param name="domainEvent">The domain event.</param>
+        /// <param name="cancellationToken">The cancellation token.</param>
+        private async Task HandleSpecificDomainEvent(TransactionHasBeenCompletedEvent domainEvent,
+                                                     CancellationToken cancellationToken)
+        {
+            TransactionAggregate transactionAggregate =
+                await this.TransactionAggregateManager.GetAggregate(domainEvent.EstateId, domainEvent.TransactionId, cancellationToken);
+
+            if (transactionAggregate.IsAuthorised == false)
+            {
+                // Ignore not successful transactions
+                return;
+            }
+
+            if (transactionAggregate.IsCompleted == false || transactionAggregate.TransactionType == TransactionType.Logon ||
+                (transactionAggregate.ContractId == Guid.Empty || transactionAggregate.ProductId == Guid.Empty))
+            {
+                // These transactions cannot have fee values calculated so skip
+                return;
+            }
+
+            this.TokenResponse = await this.GetToken(cancellationToken);
```

Comment:
Codacy found an issue: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/stuart_ferguson1/TransactionProcessor/pullRequest?prid=6000729)